dbms/src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -3807,4 +3830,88 @@ bool MergeTreeData::moveParts(CurrentlyMovingPartsTagger && moving_tagger)
     return true;
 }
 
+void MergeTreeData::loadPartsInsertionHistory()
+{
+    String full_path = getDataPaths()[0];
+    String inserted_parts_path = Poco::Path(full_path).makeAbsolute().toString() + "inserted_parts.txt";
+    Poco::File file(inserted_parts_path);
+    bool created = file.createFile();
+    if (created)
+    {
+        LOG_INFO(log, "Created " << inserted_parts_path);
+    }
+
+    // Fix corrupted file when necessary
+    size_t file_size = file.getSize();
+    if (file_size % sizeof(UInt128) != 0)
+    {
+        size_t exp_file_size = (file_size % sizeof(UInt128))*sizeof(UInt128);
+        file.setSize(exp_file_size);
+        LOG_ERROR(log, inserted_parts_path << " size " << file_size << " is incorrect. Truncated it to " << exp_file_size << ".");
+        file_size = exp_file_size;
+    }
+
+    // Shorten file when necessary
```

Comment:
Maybe just use cyclic buffer?

Reply:
I'll use a text file to store block_id (partition_id + 128bit hash).